### PR TITLE
Dotcom FSE Beta: Show opt-in step to 1% of existing users

### LINKED ISSUE #56506: FSE beta: show the opt-in only to a certain fraction of users
To avoid having too many users opting into FSE beta, we want to limit the number of users that will see the opt-in step. This feature is still in its early stages, and we want to avoid overloading support queues.

In https://github.com/Automattic/wp-calypso/issues/56290, we concluded that the Experimentation platform is not a good fit for this use case. Our fallback option is to determine eligibility based on `user_id`. E.g., we could use `user_id mod x === 0`, and lower the `x` over time to increase the percentage. We can also implement some other approaches, but it should allow us to ramp up/down the exposure quickly while retaining the setting for existing eligible users.

### PR DESCRIPTION
**Pull request by Copons**
Based on https://github.com/Automattic/wp-calypso/pull/56514.

#### Changes proposed in this Pull Request

* Show the Dotcom FSE Beta opt-in step to 1% of existing users.
* Also keep showing the step on environments with `full-site-editing/beta-opt-in` enabled.

The percentage is determined on the user ID with a simple formula, which lends itself to being quickly adjusted at will.
```js
userId % 100 < percentage
```

#### Note

There already is a [different implementation](https://github.com/Automattic/wp-calypso/blob/ce4a877e24ae50f019eb9b60156e69ebcdbad6f8/client/lib/naive-client-side-rollout/index.ts) of this "naive rollout" in Calypso.
I've asked the relevant team about it, and if it would suit our needs.

I've tried it briefly, and I've noticed that the `currentUser` object used by it is always undefined. I'll investigate it if we think reusing that function is better.

Although, I think that my super simple implementation is enough for the rollout, and it also gives us some degree of (manual) control over who gets to be eligible to the beta.

#### Testing instructions

* Make note of your user ID.
* Spin up local Calypso.
* Open http://calypso.localhost:3000/new with an existing user.
  * Make sure the second step is the FSE Beta opt-in.
* Open http://calypso.localhost:3000/new?flags=-full-site-editing/beta-opt-in (to disable the feature flag) with an exiting user.
  * If your user ID's last 2 digits are `< 10`:
    * Make sure the second step is the FSE Beta opt-in.
  * Otherwise:
    * Make sure the FSE Beta opt-in step is never shown.
    * Adjust the `FSE_BETA_ROLLOUT_PERCENTAGE` to be strictly `>` your user ID's last 2 digits.
    (e.g. user `12345` would need a percentage of `46` to be eligible to the FSE Beta)
    * Try again, and make sure this time the second step is the FSE Beta opt-in.
* Open http://calypso.localhost:3000/new in an anonymous tab (or logged out).
  * Progress through the steps.
  * Make sure the FSE Beta opt-in step is never shown.

Fixes #56506

### CODE REVIEW

**Review by Addison-Stavlo — COMMENTED** [review]
I tested and everything works as expected per testing instructions:
* Open http://calypso.localhost:3000/new with an existing user. - ✅  opt-in step shows up
* Open http://calypso.localhost:3000/new?flags=-full-site-editing/beta-opt-in  - ✅  opt-in step does not show up as expected since my userId is outside the eligible group.
* ✅  Updating the opt in % then allows the beta opt in show up in the above flow with my user ID.
* Open http://calypso.localhost:3000/new in an anonymous tab - ✅  beta opt-in is now shown

<hr/>

> There already is a different implementation of this "naive rollout" in Calypso.
I've asked the relevant team about it, and if it would suit our needs.

It looks like it should work for our needs as well, as long as the `currentUser` being `undefined` issue is resolved.

> Although, I think that my super simple implementation is enough for the rollout, and it also gives us some degree of (manual) control over who gets to be eligible to the beta.

Also agree!  If it doesn't make sense to go with the other mentioned approach, this seems good.

**Comment by Copons** [comment]
I've added the Do Not Merge label as a warning that this PR, when merged, would practically start the beta. 😄

File: client/landing/gutenboarding/hooks/use-fse-beta-eligibility.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { isEnabled } from '@automattic/calypso-config';
+import { useSelect } from '@wordpress/data';
+import { USER_STORE } from '../stores/user';
+
+const FSE_BETA_ROLLOUT_PERCENTAGE = 10;
```
Comment (vindl) [inline client/landing/gutenboarding/hooks/use-fse-beta-eligibility.ts]: Such bravery! I'd start with `1` and go from there. 😅

File: client/landing/gutenboarding/hooks/use-fse-beta-eligibility.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+export default function useFseBetaEligibility(): boolean {
+	const currentUser = useSelect( ( select ) => select( USER_STORE ).getCurrentUser() );
+
+	// FSE Beta is only active for existing users.
+	if ( ! currentUser?.ID ) {
+		return false;
+	}
+
+	// Force the FSE Beta on development environment.
```
Comment (vindl) [inline client/landing/gutenboarding/hooks/use-fse-beta-eligibility.ts]: For completeness sake, this will also force it on Horizon environment.

**Review by vindl — COMMENTED** [review]
I think there are three minor caveats to be aware of with this approach:

1. If we decide to lower the percentage (say due to high support demand), some users that previously saw the beta opt-in would no longer see it.
2. The eligible values will not be uniformly distributed, possibly causing another non-obvious clumping (e.g., consecutive new users, etc.). This shouldn't be the case with the hash-based approach linked in summary.
3. It doesn't support non-integer values (`0.1%` plx? 😄 )

Are any of these _really_ important for our use case? Most likely not. 😄

File: client/landing/gutenboarding/hooks/use-fse-beta-eligibility.ts (hunk 2 above)
Comment (Copons) [reply client/landing/gutenboarding/hooks/use-fse-beta-eligibility.ts]: Whoopsie, that's a typo for `environments` (plural), but you're right, it's just easier to be clear and mention Horizon explicitly.

File: client/landing/gutenboarding/hooks/use-fse-beta-eligibility.ts (hunk 1 above)
Comment (Copons) [reply client/landing/gutenboarding/hooks/use-fse-beta-eligibility.ts]: Why not `0.1` then? :trollface:

**Comment by Copons** [comment]
> 1. If we decide to lower the percentage (say due to high support demand), some users that previously saw the beta opt-in would no longer see it.

@vindl I don't see it as a big deal for several reasons.
The scenario you mention would affect users creating _at least_ 3 sites. I don't have data on hand, but I'd say... they are not that many? 😄 
Also, the technical effort needed for that kind of setup is most likely not worth it.

> 2. The eligible values will not be uniformly distributed, possibly causing another non-obvious clumping (e.g., consecutive new users, etc.). This shouldn't be the case with the hash-based approach linked in summary.

The clumping can happen in theory, but in practice it would mean that a number of consecutive users created new sites all together. Which is probably not happening. 😄 

> 3. It doesn't support non-integer values (`0.1%` plx? 😄 )

_You are so preoccupied with whether or not we could, that you didn't stop to think if we should._ 🦖

**Comment by vindl** [comment]
@Copons That's why I pointed out that these are _minor caveats_ that are not important now. 😄  Mostly surfacing them so we are aware of the present constraints in the future. E.g., (1) might become more relevant once we start moving the opt-in from the signup flow to the admin interface.

File: client/landing/gutenboarding/hooks/use-fse-beta-eligibility.ts (hunk 1 above)
Comment (vindl) [inline client/landing/gutenboarding/hooks/use-fse-beta-eligibility.ts]:
```suggestion
const FSE_BETA_ROLLOUT_PERCENTAGE = 1;
```

😄

**Review by vindl — APPROVED** [review]

**Comment by zaguiini** [comment]
This works.

One question, though: why didn't we just create one A/B test using Abacus for example, instead of manually managing the experiment and its implementation inside the code?

**Comment by mattwiebe** [comment]
> why didn't we just create one A/B test using Abacus for example, instead of manually managing the experiment and its implementation inside the code?

For a bunch of reasons! See paYE8P-12q-p2 for more context.

**Review by zaguiini — APPROVED** [review]

File: client/landing/gutenboarding/hooks/use-steps.ts [hunk 3]
```diff
@@ -61,8 +62,8 @@ export default function useSteps(): Array< StepType > {
 		steps = steps.filter( ( step ) => step !== Step.Style );
 	}
 
-	// Remove the FSE Beta Opt In if not in a configured environment
-	if ( ! isEnabled( 'full-site-editing/beta-opt-in' ) ) {
```
Comment (tyxla) [inline client/landing/gutenboarding/hooks/use-steps.ts:65]: Noting that after we changed this, the `isEnabled` import is unused and can be removed.

Comment (Copons) [reply client/landing/gutenboarding/hooks/use-steps.ts:65]: Yup, I noticed it while working on something unrelated.
Wonder why the commit hooks didn't ring the alarm bell though. 🤔 

I'm working on another change on this, and I'll remove the import in there, unless somebody else finds their way here sooner. 🙂

**Pull request merged**